### PR TITLE
Add firewalld utility class

### PR DESCRIPTION
**Pull request by pbrezina**

### CODE REVIEW

**Comment by pbrezina** [comment]
@ikerexxe @justin-stephenson May I ask you for a review?

This adds basic firewalld and windows firewall support. Can be used like:

```
@pytest.mark.tier(0)
@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
def test_passkey(client: Client, provider: GenericProvider):
    u = provider.user("tuser").add()
    provider.firewall.drop(389)

    client.sssd.start()
    assert client.tools.id("tuser") is None
```

File: pytest_mh/utils/firewall.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+class GenericFirewall(ABC, MultihostUtility):
+    """
+    Configure host firewall.
+
+    All changes are automatically reverted when a test is finished.
+    """
+
+    @abstractmethod
+    def accept(self, what: PortSpec | list[PortSpec]) -> None:
```
Comment (justin-stephenson) [inline pytest_mh/utils/firewall.py:32]: Can a more descriptive variable name be used other than `what` ? Same applies for all uses of `what` in this file

File: pytest_mh/utils/firewall.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        self._backup: str = "C:\\.mh_firewall.bak.wfw"
+
+    def setup_when_used(self):
+        """
+        Create a backup of current firewall configuration.
+
+        :meta private:
+        """
+        super().setup_when_used()
+        self.logger.info(f"Windows Firewall: creating backup on {self.host.hostname}")
```
Comment (justin-stephenson) [inline pytest_mh/utils/firewall.py:236]: Worth adding the backup path + filename to this log message?

File: pytest_mh/utils/firewall.py (hunk 1 above)
Comment (pbrezina) [reply pytest_mh/utils/firewall.py]: Switch to port (GenericFirewall, WindowsFirewall) and port_and_service (LinuxFirewalld)

File: pytest_mh/utils/firewall.py (hunk 2 above)
Comment (pbrezina) [reply pytest_mh/utils/firewall.py]: Done.

**Comment by pbrezina** [comment]
> There are 4 commits and only one of them is about firewalld. Is it intentional?

Yes, actually the third commit is requirement for the forth. The second commit is also required for passkey testing and working with multiple branches would not be easy. And the first commit just happened to be here but I could just merge it  :-)

**Review by justin-stephenson — APPROVED** [review]
LGTM, thanks!

**Review by ikerexxe — COMMENTED** [review]
Just two minor comments

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -10,55 +10,27 @@ concurrency:
...
-        pip3 install virtualenv
-        python3 -m venv .venv
-        source .venv/bin/activate
-        pip3 install -r ./requirements.txt
+        sudo apt-get update
 
-    - name: Run tests
-      run: source .venv/bin/activate && pytest --color=yes -vv
+        python -m pip install --upgrade pip
+        pip install tox tox-gh
```
[the remark this answers is not in the document]
Comment (pbrezina) [reply .github/workflows/ci.yml:31]: Actually apt-get update is not needed.

File: pytest_mh/ssh.py [hunk 4]
```diff
@@ -902,18 +902,21 @@ def exec(
             raise_on_error=raise_on_error,
         )
 
-    def expect(self, expect_script: str) -> SSHProcessResult:
+    def expect(self, expect_script: str, raise_on_error: bool = False) -> SSHProcessResult:
```
[the remark this answers is not in the document]
Comment (pbrezina) [reply pytest_mh/ssh.py:905]: Done.

**Review by ikerexxe — APPROVED** [review]
LGTM!

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -10,55 +10,25 @@ concurrency:
...
-        python3 -m venv .venv
-        source .venv/bin/activate
-        pip3 install -r ./requirements.txt
-
-    - name: Run tests
-      run: source .venv/bin/activate && pytest --color=yes -vv
+        python -m pip install --upgrade pip
+        pip install tox tox-gh
+    - name: Run tox
+      run: tox --colored=yes
```
Comment (webknjaz) [inline .github/workflows/ci.yml:31]: @pbrezina FYI it's customary to split the env provisioning and testing into two commands so that it's visible in the CI which step fails. This example shows how to do this: https://github.com/tox-dev/tox-gh#workflow-configuration.

Also, here's a neat trick to re-run only the tests that failed with maximum verbosity, keeping the job status failing: https://github.com/cherrypy/cheroot/blob/0fd16f0/.github/workflows/ci-cd.yml#L880-L895. I thought you might find it interesting.

Comment (pbrezina) [reply .github/workflows/ci.yml:31]: Cool, thank you.
